### ISSUE TITLE
interop with go-libp2p: Multiplex

### ISSUE THREAD

**Issue by diasdavid** [ipfs/js-libp2p-ipfs-nodejs#59]
> Following the notes on: https://github.com/ipfs/js-libp2p-ipfs-nodejs/pull/55

We've been observing a weird behavior when using multiplex to do interop between `js-libp2p-ipfs-nodejs` and `go-libp2p`, connections being closed or data being capped. We are still investigating these problems.

### set up

Assuming that you have golang and Node.js installed.

1. Clone go-libp2p repo and install deps

```sh
> mkdirp $GOPATH/src/github.com/libp2p
> cd $GOPATH/src/github.com/libp2p
> git clone github.com:libp2p/go-libp2p.git
> make && make deps
```

2. Change go-libp2p dep, go-libp2p-swarm, to use multiplex

```sh
> cd $GOPATH/src/github.com/libp2p/go-libp2p-swarm
# Checkout this branch https://github.com/libp2p/go-libp2p-swarm/pull/16
> git checkout add-multiplex
```

3. Build the echo example of go-libp2p

```sh
> cd $GOPATH/src/github.com/libp2p/go-libp2p/examples/echo
> go build main.go
```

### echo more than 1MB from js to go

`Writting it`

### echo more than 1MB from go to js

`Writting it`

-----------------------------------------

## @whyrusleeping's new tool - `p2pc` (p2p netcat)

- https://github.com/whyrusleeping/p2pc

**Comment by whyrusleeping** [comment]
Guys, the echo program is just a demo. It echos a single message of at most 1024 bytes once: https://github.com/libp2p/go-libp2p/blob/master/examples/echo/main.go#L147

**Comment by diasdavid** [comment]
Thanks for the simple echo tool @whyrusleeping https://github.com/whyrusleeping/p2pc

Just ran with SPDY and it is all good, but with multiplex, it breaks in the JS side:

![image](https://cloud.githubusercontent.com/assets/1211152/22263252/b304670a-e26b-11e6-8120-8fa0d72f8676.png)

**Comment by whyrusleeping** [comment]
@diasdavid you can enable secio on that tool by setting the `-secio` flag

**Comment by dignifiedquire** [comment]
I found that

- Using `pull.values` can result in an issue where the multiplex stream is closed before the go side has time to reply. This results in the "cannot write to closed stream" error. Using `pull-pushable` as source and ending in a timeout fixes that, but is less than ideal. This seems like a bug on our side, as we should not close things before we don't have the ack on the mulitplex stream for the close.
- Echo with small values generally works, but the `conn` that we get from the dial is not ended. This is not a problem with multiplex it ends it stream and also the direct wrapper to pull-streams is ended. But the `Connection` that is propagated is not ended. I suspect the issue to be in multistream-select, which is the layer between the two.

**Comment by diasdavid** [comment]
> Using pull.values can result in an issue where the multiplex stream is closed before the go side has time to reply.

It should just `half` close, no? Seems like a error indeed and a quick test that can be added to libp2p-multiplex //cc @VictorBjelkholm.

> I suspect the issue to be in multistream-select, which is the layer between the two.

@VictorBjelkholm can you finish this PR -- https://github.com/multiformats/js-multistream-select/pull/34 -- and see if this issue is coming from multistream-select?

**Comment by diasdavid** [comment]
> Using pull.values can result in an issue where the multiplex stream is closed before the go side has time to reply.

What @dignifiedquire is suggesting is that opening and closing a muxed stream on top of multiplex, will also close the whole muxer if no other stream was open and since the whole muxer is closed, when go tries to write back, the socket was already closed.

**Comment by diasdavid** [comment]
@whyrusleeping here is your repro case

1. git clone this repo
2. npm install
3. cd examples/echo
4. spawn your p2pc as listener
5. LIBP2P_MUXER=multiplex node dialer.js

**Comment by whyrusleeping** [comment]
Just a note, i read something a few comments up about 'half closing'. What do you guys mean there?

**Comment by diasdavid** [comment]
@whyrusleeping half closing means that we send the FIN packet but never receive it from the other side, so what happens is that you can no longer write, but until we receive the same FIN packet, we will still have the stream open to read data.

**Comment by diasdavid** [comment]
Updated `diasdavid/multiplex` with the bug fixes of Monday

Published libp2p-multiplex, verified that passes all the tests

Updated this module, but seeing now an issue in this test - https://travis-ci.org/ipfs/js-libp2p-ipfs-nodejs/jobs/201784327#L2170-L2182

Tried js-ipfs with it anyway and not having success. @whyrusleeping could you double check to see if this was the correct fix we did: https://github.com/diasdavid/multiplex/pull/2 I might have missed something.

**Comment by dignifiedquire** [comment]
Fixed with libp2p-multiplex@0.4.1
